File: ta4j-core/src/main/java/org/ta4j/core/indicators/RWILowIndicator.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators;

import org.ta4j.core.BarSeries;
import org.ta4j.core.num.NaN;
import org.ta4j.core.num.Num;

/**
 * The Class RandomWalkIndexLowIndicator.
 *
 * @see <a href=
 *      "http://https://rtmath.net/helpFinAnalysis/html/934563a8-9171-42d2-8444-486691234b1d.html">Source
 *      of formular</a>
 */
public class RWILowIndicator extends CachedIndicator<Num> {

    private final int barCount;
    private final transient ATRIndicator[] atrIndicators;
    private final transient Num[] sqrtValues;

    /**
     * Constructor.
     *
     * @param series   the bar series
     * @param barCount the time frame
     */
    public RWILowIndicator(BarSeries series, int barCount) {
        super(series);
        this.barCount = barCount;
        int helperCount = Math.max(0, barCount + 1);
        this.atrIndicators = new ATRIndicator[helperCount];
        this.sqrtValues = new Num[helperCount];
        for (int n = 2; n <= barCount; n++) {
            atrIndicators[n] = new ATRIndicator(series, n);
            sqrtValues[n] = series.numFactory().numOf(n).sqrt();
        }
    }

    @Override
    protected Num calculate(int index) {
        if (index < getCountOfUnstableBars() || index - barCount + 1 < getBarSeries().getBeginIndex()) {
            return NaN.NaN;
        }

        Num maxRWIL = getBarSeries().numFactory().zero();
        boolean hasDefinedValue = false;
        for (int n = 2; n <= barCount; n++) {
            Num rwiLow = calcRWILFor(index, n);
            if (!IndicatorUtils.isInvalid(rwiLow)) {
                maxRWIL = hasDefinedValue ? maxRWIL.max(rwiLow) : rwiLow;
                hasDefinedValue = true;
            }
        }

        return hasDefinedValue ? maxRWIL : NaN.NaN;
    }

    @Override
    public int getCountOfUnstableBars() {
        return Math.max(0, barCount);
    }

    private Num calcRWILFor(final int index, final int n) {
        BarSeries series = getBarSeries();
        Num low = series.getBar(index).getLowPrice();
        Num highN = series.getBar(index + 1 - n).getHighPrice();
        Num atrN = atrIndicators[n].getValue(index);
        Num sqrtN = sqrtValues[n];

        return highN.minus(low).dividedBy(atrN.multipliedBy(sqrtN));
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + " barCount: " + barCount;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/indicators/RWILowIndicator.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators

import org.ta4j.core.BarSeries
import org.ta4j.core.num.NaN
import org.ta4j.core.num.*

/**
 * The Class RandomWalkIndexLowIndicator.
 *
 * @see [Source
 * of formular](http://https://rtmath.net/helpFinAnalysis/html/934563a8-9171-42d2-8444-486691234b1d.html)
 */
class RWILowIndicator
/**
 * Constructor.
 *
 * @param series   the series
 * @param barCount the time frame
 */(series: BarSeries?, private val barCount: Int) : CachedIndicator<Num>(series) {
    override fun calculate(index: Int): Num {
        if (index - barCount + 1 < barSeries!!.beginIndex) {
            return NaN.NaN
        }
        var minRWIL = numOf(0)
        for (n in 2..barCount) {
            minRWIL = minRWIL.max(calcRWIHFor(index, n))
        }
        return minRWIL
    }

    private fun calcRWIHFor(index: Int, n: Int): Num {
        val series = barSeries
        val low = series!!.getBar(index).lowPrice!!
        val highN = series.getBar(index + 1 - n).highPrice
        val atrN = ATRIndicator(series, n)[index]
        val sqrtN = numOf(n).sqrt()
        return highN!!.minus(low).div(atrN.times(sqrtN))
    }

    override fun toString(): String {
        return javaClass.simpleName + " barCount: " + barCount
    }
}